google-cloud-spanner/src/main/java/com/google/cloud/spanner/GrpcStruct.java

```diff
@@ -355,64 +396,96 @@ protected <T extends ProtocolMessageEnum> T getProtoEnumInternal(
 
   @Override
   protected boolean getBooleanInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (Boolean) rowData.get(columnIndex);
   }
 
   @Override
   protected long getLongInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (Long) rowData.get(columnIndex);
   }
 
   @Override
   protected double getDoubleInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (Double) rowData.get(columnIndex);
   }
 
   @Override
   protected BigDecimal getBigDecimalInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (BigDecimal) rowData.get(columnIndex);
   }
 
   @Override
   protected String getStringInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (String) rowData.get(columnIndex);
   }
 
   @Override
   protected String getJsonInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (String) rowData.get(columnIndex);
   }
 
   @Override
   protected String getPgJsonbInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (String) rowData.get(columnIndex);
   }
 
   @Override
   protected ByteArray getBytesInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return getLazyBytesInternal(columnIndex).getByteArray();
   }
 
   LazyByteArray getLazyBytesInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (LazyByteArray) rowData.get(columnIndex);
   }
 
   @Override
   protected Timestamp getTimestampInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (Timestamp) rowData.get(columnIndex);
   }
 
   @Override
   protected Date getDateInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
     return (Date) rowData.get(columnIndex);
   }
 
   protected com.google.protobuf.Value getProtoValueInternal(int columnIndex) {
     return (com.google.protobuf.Value) rowData.get(columnIndex);
   }
 
+  private void ensureDecoded(int columnIndex) {
+    if (decodeMode == DecodeMode.LAZY_PER_ROW && !rowDecoded) {
+      for (int i = 0; i < rowData.size(); i++) {
+        rowData.set(
+            i,
+            decodeValue(
+                type.getStructFields().get(i).getType(),
+                (com.google.protobuf.Value) rowData.get(i)));
+      }
+      rowDecoded = true;
+    } else if (decodeMode == DecodeMode.LAZY_PER_COL && !colDecoded.get(columnIndex)) {
+      rowData.set(
+          columnIndex,
+          decodeValue(
+              type.getStructFields().get(columnIndex).getType(),
+              (com.google.protobuf.Value) rowData.get(columnIndex)));
+      colDecoded.set(columnIndex);
+    }
+  }
+
   @Override
   protected Value getValueInternal(int columnIndex) {
+    ensureDecoded(columnIndex);
```

Comment:
If DecodeMode is not direct then cant we just directly return the Value without decoding?

Reply:
I don't think so, but I'm also not sure whether there might be a misunderstanding here or not:
1. Here we are returning a `com.google.cloud.spanner.Value`. So a decoded Value instance as defined by the Spanner library, and not a protobuf value.
2. The underlying value, when decode mode is not DIRECT, could be a `com.google.protobuf.Value`. We need to decode that value into the corresponding Java object when we want to return a `com.google.cloud.spanner.Value` instance.

Reply:
Understood, missed that we are returning a `com.google.cloud.spanner.Value`